File: error_analysis/entailment/classifier.clj
(ns classifier
  "Classification script to demonstrate how the entailment data is biased
  because of how it's constructed.

  When sentence1 is a substring of sentence2, the label is always 'ENTAILMENT'.
  For 'CONTRADICTION' and 'NEUTRAL', it's not as obvious."
  {:clj-kondo/ignore [:unresolved-namespace]}
  (:require [clojure.string :as str]
            [cheshire.core :as json]
            [clojure.java.io :as io]))

(defn levenshtein-distance
  "Levenshtein distance between two strings."
  [str1 str2]
  (let [m (count str1)
        n (count str2)
        dp (make-array Integer/TYPE (inc m) (inc n))]

    (dotimes [i (inc m)]
      (aset-int dp i 0 i))

    (dotimes [j (inc n)]
      (aset-int dp 0 j j))

    (dotimes [i m]
      (dotimes [j n]
        (let [cost (if (= (nth str1 i) (nth str2 j)) 0 1)]
          (aset-int dp (inc i) (inc j)
                    (min (+ (aget dp i j) cost)
                         (+ 1 (aget dp (inc i) j))
                         (+ 1 (aget dp i (inc j))))))))

    (aget dp m n)))

#_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
(defn levenshtein-similarity
  "Similarity between two strings using the Levenshtein distance."
  [str1 str2]
  (let [len-sum (+ (count str1) (count str2))
        dist (levenshtein-distance str1 str2)]
    (float (- 1 (/ dist len-sum)))))

;; BAG OF WORDS
(defn set->map
  "Set and create a map from each element to an index."
  [s]
  (into {} (map vector s (range))))

(defn word-vector
  "Helper function for bag-of-words. Creates a vector of size |vocab|
  with the count of each word in words."
  [vocab words]
  (reduce (fn [acc word]
            (let [index (vocab word)]
              (update acc index inc)))
          (vec (repeat (count vocab) 0))
          words))

(defn split-words
  "Split a string on whitespace."
  [s]
  (clojure.string/split s #"\s+"))

(defn bag-of-words
  "Create a bag of words representation for two strings.
  The bag of words representation is a vector of size |vocab| where
  vocab is the union of unique of words between the strings."
  [vocab s]
  (word-vector vocab (split-words s)))

(defn make-vocab
  "Create a vocabulary from a collection of strings."
  [coll]
  (set->map (set (mapcat #(split-words %) coll))))
;; END OF BAG OF WORDS

(defn jaccard
  "Jaccard similarity between two vectors."
  [vector1 vector2]
  (let [sum-min (reduce + (map min vector1 vector2))
        sum-max (reduce + (map max vector1 vector2))]
    (float (/ sum-min sum-max))))

(defn norm
  "2-norm of a vector."
  [vector]
  (Math/sqrt (reduce + (map #(* % %) vector))))

(defn cosine
  "Cosine similarity between two vectors."
  [vector1 vector2]
  (let [dot-product (reduce + (map * vector1 vector2))
        norm1 (norm vector1)
        norm2 (norm vector2)]
    (float (/ dot-product (* norm1 norm2)))))

(defn jaccard-similarity
  "Jaccard similarity between two strings using the Generalised Jaccard index."
  [str1 str2]
  (let [vocab (make-vocab [str1 str2])]
    (jaccard (bag-of-words vocab str1)
             (bag-of-words vocab str2))))

#_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
(defn cosine-similarity
  "Cosine similarity between two strings."
  [str1 str2]
  (let [vocab (make-vocab [str1 str2])]
    (cosine (bag-of-words vocab str1)
            (bag-of-words vocab str2))))

(def similarity-threshold 0.75)

(defn similar?
  "Returns true if the (sim-fn str1 str2) is greater
  than or equal to the similarity threshold."
  [str1 str2 sim-fn threshold]
  (>= (sim-fn str1 str2) threshold))

(defn symmetric-substring
  "Returns true if str1 is a substring of str2 or vice versa."
  [str1 str2]
  (or (str/includes? str1 str2) (str/includes? str2 str1)))

(defn classify
  "Classify two strings as entailment, contradiction or neutral."
  ([str1 str2 sim-fn threshold]
   (cond
     (symmetric-substring str1 str2) "ENTAILMENT"
     (similar? str1 str2 sim-fn threshold) "CONTRADICTION"
     :else "NEUTRAL")))

(defn classify*
  "Classify a collection of maps with :sentence1 and :sentence2 keys."
  [coll sim-fn threshold]
  (map #(classify (:sentence1 %) (:sentence2 %) sim-fn threshold)
       coll))

(defn calc-accuracy
  "Calculate the accuracy of the classifier on a collection of maps with
  :sentence1, :sentence2 and :label keys."
  [coll sim-fn threshold]
  (let [gold (map :label coll)
        pred (classify* coll sim-fn threshold)
        correct (filter #(apply = %) (map vector gold pred))]
    (float (/ (count correct) (count coll)))))

(defn get-sim-fn
  "Get the similarity function from a string."
  [s]
  (case s
    "jaccard" jaccard-similarity
    "cosine" cosine-similarity
    "levenshtein" levenshtein-similarity
    (throw (Exception. (str "Unknown similarity function: " s)))))

(def opt-spec
  {:coerce {:max-samples :long
            :threshold :double}
   :alias {:m :max-samples}
   :exec-args {:similarity "jaccard"
               :threshold similarity-threshold}
   :args->opts [:data-file]})

(let [opts (babashka.cli/parse-opts *command-line-args* opt-spec)
      data-file (:data-file opts)
      sim-fn (get-sim-fn (:similarity opts))
      max-samples (:max-samples opts)
      threshold (:threshold opts)]
  (when (empty? data-file)
    (println "Usage: bb levenshtein.clj <data-file>")
    (System/exit 1))

  (let [data (json/parse-stream (io/reader data-file) true)
        n (or max-samples (count data))
        score (calc-accuracy (take n data) sim-fn threshold)]
    (println "Accuracy" (format "%.2f%%" (* 100 score)))))

(let [str1 "hey there man hey"
      str2 "hey there girl there there"
      vocab (make-vocab [str1 str2])
      bow1 (bag-of-words vocab str1)
      bow2 (bag-of-words vocab str2)]
  (println bow1 "=>" str1)
  (println bow2 "=>" str2))

File: error_analysis/entailment/classifier.py
import json
import sys
from collections.abc import Callable
from pathlib import Path
from typing import Optional

import typer


def levensthein_distance(str1: str, str2: str) -> int:
    m, n = len(str1), len(str2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        dp[i][0] = i

    for j in range(n + 1):
        dp[0][j] = j

    for i, c1 in enumerate(str1, 1):
        for j, c2 in enumerate(str2, 1):
            dp[i][j] = min(
                dp[i - 1][j - 1] + (c1 != c2),
                dp[i - 1][j] + 1,
                dp[i][j - 1] + 1,
            )

    return dp[m][n]


def levenshtein_similarity(str1: str, str2: str) -> float:
    len_sum = len(str1) + len(str2)
    dist = levensthein_distance(str1, str2)
    return 1 - dist / len_sum


def set_to_map(s: set[str]) -> dict[str, int]:
    return {k: i for i, k in enumerate(s)}


def word_vector(vocab: dict[str, int], words: list[str]) -> list[int]:
    vec = [0] * len(vocab)
    for w in words:
        if w in vocab:
            vec[vocab[w]] += 1
    return vec


def bag_of_words(vocab: dict[str, int], s: str) -> list[int]:
    return word_vector(vocab, s.split())


def make_vocab(corpus: list[str]) -> dict[str, int]:
    vocab: set[str] = set()
    for s in corpus:
        vocab.update(s.split())
    return set_to_map(vocab)


def jaccard(vector1: list[int], vector2: list[int]) -> float:
    sum_min = sum(min(v1, v2) for v1, v2 in zip(vector1, vector2))
    sum_max = sum(max(v1, v2) for v1, v2 in zip(vector1, vector2))
    return sum_min / sum_max


def norm(vector: list[int]) -> float:
    return sum(v**2 for v in vector) ** 0.5


def cosine(vector1: list[int], vector2: list[int]) -> float:
    dot_product = sum(v1 * v2 for v1, v2 in zip(vector1, vector2))
    return dot_product / (norm(vector1) * norm(vector2))


def jaccard_similarity(str1: str, str2: str) -> float:
    vocab = make_vocab([str1, str2])
    vec1 = bag_of_words(vocab, str1)
    vec2 = bag_of_words(vocab, str2)
    return jaccard(vec1, vec2)


def cosine_similarity(str1: str, str2: str) -> float:
    vocab = make_vocab([str1, str2])
    vec1 = bag_of_words(vocab, str1)
    vec2 = bag_of_words(vocab, str2)
    return cosine(vec1, vec2)


SIMILARITY_THRESHOLD = 0.75


def is_similar(
    str1: str, str2: str, sim_func: Callable[[str, str], float], threshold: float
) -> bool:
    return sim_func(str1, str2) >= threshold


def is_symmetric_substring(str1: str, str2: str) -> bool:
    return str1 in str2 or str2 in str1


def classify_pair(
    str1: str, str2: str, sim_func: Callable[[str, str], float], threshold: float
) -> str:
    if is_symmetric_substring(str1, str2):
        return "ENTAILMENT"
    if is_similar(str1, str2, sim_func, threshold):
        return "CONTRADICTION"
    return "NEUTRAL"


def classifiy(
    data: list[dict[str, str]], sim_func: Callable[[str, str], float], threshold: float
) -> list[str]:
    return [
        classify_pair(d["sentence1"], d["sentence2"], sim_func, threshold) for d in data
    ]


def calc_accuracy(
    data: list[dict[str, str]], sim_func: Callable[[str, str], float], threshold: float
) -> float:
    gold = [d["label"] for d in data]
    pred = classifiy(data, sim_func, threshold)
    correct = sum(g == p for g, p in zip(gold, pred))
    return correct / len(gold)


def get_sim_func(s: str) -> Callable[[str, str], float]:
    if s == "jaccard":
        return jaccard_similarity
    if s == "cosine":
        return cosine_similarity
    if s == "levenshtein":
        return levenshtein_similarity
    raise ValueError(f"Unknown similarity function: {s}")


def main(
    file: Path,
    similarity: str = "jaccard",
    max_samples: Optional[int] = None,
    threshold: float = SIMILARITY_THRESHOLD,
) -> None:
    data = json.loads(file.read_text())[:max_samples]
    score = calc_accuracy(data, get_sim_func(similarity), threshold)
    print(f"Accuracy: {score:.2%}")


if __name__ == "__main__":
    if not hasattr(sys, "ps1"):  # not in interactive mode
        typer.run(main)


def rcf():
    "..."
    str1 = "hey there man hey"
    str2 = "hey there girl there there"
    vocab = make_vocab([str1, str2])
    bow1 = bag_of_words(vocab, str1)
    bow2 = bag_of_words(vocab, str2)
    print(bow1, "->", str1)
    print(bow2, "->", str2)
